bigchaindb/tep/upsert_validator.py

```diff
@@ -0,0 +1,216 @@
+            'name': self.name,
+            'version': self.version,
+            'args': args
+        }
+        return new_election_object
+
+    def _recipients(self):
+        """Convert validator dictionary to a recipient list for `Transaction`"""
+
+        recipients = []
+        for public_key, voting_power in self._validators().items():
+            recipients.append(([public_key], voting_power))
+
+        return recipients
+
+    def _validators(self):
+        """Return a dictionary of validators with key as `public_key` and
+           value as the `voting_power`
+        """
+
+        validators = {}
+        for validator in self.bigchain.get_validators():
+            if validator['pub_key']['type'] == GO_AMINO_ED25519:
+                public_key = public_key_from_ed25519_key(key_from_base64(validator['pub_key']['value']))
+                validators[public_key] = validator['voting_power']
+            else:
+                # TODO: use appropriate exception
+                raise NotImplementedError
+
+        return validators
+
+    def _get_vote(self, tx_election, public_key):
+        """For a given `public_key` return the `input` and `amount` as
+           voting power
+        """
+        for i, tx_input in enumerate(tx_election.to_inputs()):
+            if tx_input.owners_before == [public_key] and \
+               tx_election.outputs[i].public_keys == [public_key]:
+                return ([tx_input], tx_election.outputs[i].amount)
+
+        return ([], 0)
+
+    def _is_same_topology(self, current_topology, election_topology):
+
+        voters = {}
+        for voter in election_topology:
+            [public_key] = voter.public_keys
+            voting_power = voter.amount
+            voters[public_key] = voting_power
+
+        # Check whether the voters and their votes is same to that of the
+        # validators and their voting power in the network
+        return (current_topology.items() == voters.items())
+
+    @classmethod
+    def election_id(cls, tx_election_id):
+        return base58.b58encode(bytes.fromhex(tx_election_id))
+
+    def _execute_action(self, tx):
+        (code, msg) = self.bigchain.write_transaction(tx, 'broadcast_tx_commit')
+        return (True if code == 202 else False)
+
+    def propose(self, validator_data, node_key_path):
+        validator_public_key = validator_data['public_key']
+        validator_power = validator_data['power']
+        validator_node_id = validator_data['node_id']
+
+        node_key = load_node_key(node_key_path)
+
+        args = {
+            'public_key': validator_public_key,
+            'power': validator_power,
+            'node_id': validator_node_id
+        }
+
+        tx = Transaction.create([node_key.public_key],
+                                self._recipients(),
+                                asset=self._new_election_object(args))\
+                        .sign([node_key.private_key])
+        return self._execute_action(tx)
```

Comment:
Can we make propose return actual error details? It should be quite helpful.

Reply:
In the PR you mentioned you are returning different error messages although not very specific yet. The HTTP API reports the error message it gets from `write_transaction` as well - why not returning the message here?

Reply:
Pleae have a look [here](https://github.com/bigchaindb/bigchaindb/blob/master/bigchaindb/tendermint/lib.py#L67). The error response details contain information regarding the internal exceptions that Tendermint faced when executing the transaction i.e. whenever the json rpc api returns `500` the response is usually some internal exception within Tendermint with a sepecific error code in the json response. IMHO returning these internal errors to the admin might not be of much help but if you think it might be of some use then I can return the response.

Reply:
I would return the error message from the start so that if we expose this functionality in the HTTP API or in the command line we do not have to do backwards-incompatible changes in the future when the error handling is improved.